src/current/v22.1/bulk-delete-data.md

```diff
@@ -76,22 +98,185 @@ while True:
+                  () -> new IllegalArgumentException("JDBC_DATABASE_URL is not set.")));
+            connection = ds.getConnection();
+            // call the method to perform the deletes
+            deleteData(connection);
+        } catch (SQLException e) {
+            e.printStackTrace();
+        } finally {
+            // close the connection
+            if (connection != null) connection.close();
+        }
+    }
+}
+~~~
+
+</section>
+
+<section class="filter-content" markdown="1" data-scope="csharp">
+
+A simple C# Npgsql application that bulk deletes rows in batches of 5000 would look like this:
+
+{% include_cached copy-clipboard.html %}
+~~~ csharp
+using System;
+using System.Data;
+using System.Net.Security;
+using Npgsql;
+
+namespace Cockroach
+{
+  class MainClass
+  {
+    static void Main(string[] args)
+    {
+      // create the connection string from the connection parameters
+      var connStringBuilder = new NpgsqlConnectionStringBuilder();
+      connStringBuilder.Host = "cluster-name-743.g95.cockroachlabs.cloud";
+      connStringBuilder.Port = 26257;
+      connStringBuilder.SslMode = SslMode.VerifyFull;
+      connStringBuilder.Username = "maxroach";
+      connStringBuilder.Password = "notAGoodPassword";
+      connStringBuilder.Database = "tpcc";
+      connStringBuilder.ApplicationName = "docs_bulk_delete_csharp";
+      // call the method to perform the deletes
+      DeleteRows(connStringBuilder.ConnectionString);
+    }
+
+    static void DeleteRows(string connString)
+    {
+      // create the data source with the connection string
+      using var dataSource = NpgsqlDataSource.Create(connString);
+      {
+        using var connection = dataSource.OpenConnection();
+        bool cont = true;
+        // the initial warehouse ID we will delete orders from
+        int warehouseId = 4;
+        do {
+          Console.WriteLine("Deleting data from warehouse <= to " + warehouseId);
+          using var cmd = new NpgsqlCommand("DELETE FROM new_order WHERE no_w_id <= (@p1) ORDER BY no_w_id DESC LIMIT 5000 RETURNING no_w_id", connection)
+          {
+            Parameters = 
+            {
+              // using a prepared statement and the current warehouse ID
+              new("p1", warehouseId)
+            }
+          };
+          using (var reader = cmd.ExecuteReader())
+          {
+            if (reader.HasRows)
+            {
+              while (reader.Read())
+              {
+                // Get the warehouse ID. When the result set is empty this will be
+                // the warehouse ID of the final row of this batch.
+                warehouseId = reader.GetInt32(0);
+              }
+              Console.WriteLine("Warehouse ID is now " + warehouseId); 
+            }
+            else {
+              // All the rows have been deleted, so break out of the loop
+              cont = false;
+            }
+            reader.Close();
```

Comment:
`reader` (and `connection` below) should be automatically closed https://stackoverflow.com/questions/5243398/will-a-using-block-close-a-database-connection

Reply:
Good catch. Removing.